prismicio/slice-machine
# fix(init): limit the amount of custom types created concurrently (DT-1758)

## Context

Since https://github.com/prismicio/slice-machine/releases/tag/slice-machine-ui%401.17.0, Slice Machine init command downloads remote custom types during the init process so that users can work on them locally.

When users have many custom types (60+), the init command can fail to allocate memory as it was creating all custom types concurrently which is a costly operation.

Resolves: DT-1758

## The Solution

This PR limits the rate at which custom types are created by the init command so that no more than 8 custom types are being created at the same time.

## How to QA

1. Create a repository with many custom types (60+) or ask me for a clone of such a repository
2. Try running the current init command inside a project, it **should** fail
```bash
npx create-next-app@latest
npx @slicemachine/init@latest
```
3. Try running the alpha init command, it **should** succeed
```bash
npx create-next-app@latest
npx @slicemachine/init@alpha
```

## Impact / Dependencies

/

## Checklist before requesting a review
- [ ] I hereby declare my code ready for review.
- [ ] If it is a critical feature, I have added tests.
- [ ] The CI is successful.
- [ ] If there could backward compatibility issues, it has been discussed and planned.
